### PR TITLE
Add JsonParseOptions to ignore unknown fields

### PR DESCRIPTION
**Pull request by lizan**
- add JsonParseOptions for JsonToBinaryString allow unknown fields
- rename current JsonOptions to JsonPrintOptions

### CODE REVIEW

File: src/google/protobuf/util/json_util.cc [hunk 1]
```diff
@@ -105,13 +105,17 @@ util::Status BinaryToJsonString(TypeResolver* resolver,
...
-  StatusErrorListener() : status_(util::Status::OK) {}
+  StatusErrorListener(const JsonParseOptions& options)
+      : options_(options), status_(util::Status::OK) {}
   virtual ~StatusErrorListener() {}
 
   util::Status GetStatus() { return status_; }
 
   virtual void InvalidName(const converter::LocationTrackerInterface& loc,
                            StringPiece unknown_name, StringPiece message) {
+    if (options_.ignore_unknown_fields) {
```
Comment (xfxyjwf) [inline src/google/protobuf/util/json_util.cc]: This seems the wrong place to ignore unknown fields. When options.ignore_unknown_fields is enabled, this InvalidName() method shouldn't be called at all. Could you change the JSON parsing code to not call InvalidName?

File: src/google/protobuf/util/json_util.h [hunk 2]
```diff
@@ -54,11 +61,14 @@ struct JsonOptions {
...
 
-  JsonOptions() : add_whitespace(false),
-                  always_print_primitive_fields(false) {
+  JsonPrintOptions() : add_whitespace(false),
+                       always_print_primitive_fields(false) {
   }
 };
 
+// DEPRECATED. Use JsonPrintOptions instead.
+using JsonOptions = JsonPrintOptions;
```
Comment (xfxyjwf) [inline src/google/protobuf/util/json_util.h]: Use typedef here (no C++11 features).

**Comment by anandolee** [comment]
Ignore unknown fields is disallowed in text format:
https://sites.google.com/a/google.com/protocol-buffers/migration/proto2-faq#How_do_I_ignore_unknown_fields_w

We may want to consider if JSON format should follow it too @xfxyjwf ?

**Comment by martinmaly** [comment]
There is a difference between text format and JSON in that text format is not intended for production use. JSON is and as such it can encounter data serialized by different version of the schema which could contain additional fields.

JSON serialized format is actually much like binary serialized proto in production requirements in terms of backwards/forward compatibility.

**Comment by xfxyjwf** [comment]
Unlike text format, JSON format is designed to be used as a wire format where forward compatibility (tolerance of new fields) is important. For this reason we should support ignoring unknown fields in JSON.

File: src/google/protobuf/util/json_util.h (hunk 2 above)
Comment (lizan) [inline src/google/protobuf/util/json_util.h]: Done

File: src/google/protobuf/util/json_util.cc (hunk 1 above)
Comment (lizan) [inline src/google/protobuf/util/json_util.cc]: Done

File: src/google/protobuf/util/internal/proto_writer.h [hunk 3]
```diff
@@ -76,7 +76,8 @@ class LIBPROTOBUF_EXPORT ProtoWriter : public StructuredObjectWriter {
  public:
 // Constructor. Does not take ownership of any parameter passed in.
   ProtoWriter(TypeResolver* type_resolver, const google::protobuf::Type& type,
-              strings::ByteSink* output, ErrorListener* listener);
+              strings::ByteSink* output, ErrorListener* listener,
+              bool ignore_unknown_fields = false);
```
Comment (xfxyjwf) [inline src/google/protobuf/util/internal/proto_writer.h]: Use a setter instead? We may add more flags in the future and it's not very nice to put all flags in the constructor.

Comment (lizan) [inline src/google/protobuf/util/internal/proto_writer.h]: Done

**Comment by lizan** [comment]
ping. @xfxyjwf

File: src/google/protobuf/util/internal/proto_writer.h [hunk 4]
```diff
@@ -240,7 +244,8 @@ class LIBPROTOBUF_EXPORT ProtoWriter : public StructuredObjectWriter {
 
   // Lookup the field in the current element. Looks in the base descriptor
   // and in any extension. This will report an error if the field cannot be
-  // found or if multiple matching extensions are found.
+  // found when report_unknown_names is true or if multiple matching extensions
```
Comment (xfxyjwf) [inline src/google/protobuf/util/internal/proto_writer.h]: => "ignore_unknown_names_ is false"

**Comment by xfxyjwf** [comment]
LGTM

Thanks. Please squash these commits into one commit and I'll merge it.

**Comment by lizan** [comment]
Squashed. Thanks for reviewing!

**Comment by xfxyjwf** [comment]
ok to test

**Pull request merged**